docs/source/notes/distributed_autograd.rst

```diff
@@ -0,0 +1,278 @@
+  e = rpc.rpc_sync("worker1", torch.mul, args=(b, c))
+  loss = d.sum()
+
+The associated autograd graph for the code above would be:
+
+.. image:: images/distributed_dependencies.png
+
+Computing dependencies of this distributed autograd graph is much more 
+challenging and requires some overhead (either in terms of computation or 
+network communication). 
+
+For performance sensitive applications we can avoid a 
+lot of overhead by assuming every ``send`` and ``recv`` function are valid as 
+part of the backward pass (most applications don't perform RPCs that aren't 
+used). This simplifies the distributed autograd algorithm and is much more 
+efficient, but at the cost that the application needs to be aware of the 
+limitations. This algorithm is called the `FAST mode algorithm`_ and is 
+described in detail below.
+
+In the general case it might not be necessary that every ``send`` and ``recv`` 
+function is valid as part of the backward pass. To address this, we also have 
+a `SMART mode algorithm`_ which is described in a later section.
+
+FAST mode algorithm
+-------------------
+
+The key assumption of this algorithm is that each ``send`` function has a 
+dependency of 1 when we run a backward pass. In other words, we assume we'll 
+receive a gradient over RPC from another node.
+
+The algorithm is as follows:
+
+1. We start from the worker which has the roots for the backward pass 
+   (all roots must be local).
+2. Lookup all the ``send`` functions for the current 
+   `Distributed Autograd Context`_.
+3. Compute dependencies locally starting from the provided roots and all the 
+   ``send`` functions we retrieved.
+4. After computing dependencies, kick off the local autograd engine with the 
+   provided roots.
+5. When the autograd engine executes the ``recv`` function, the ``recv`` 
+   function sends the input gradients over the wire to the appropriate worker. 
+   Each ``recv`` function knows the destination worker id since it is recorded 
+   as part of the forward pass. The ``recv`` function also sends over the 
+   ``autograd_context_id`` and ``autograd_message_id`` to the remote host.
+6. When this request is received on the remote host, we use the 
+   ``autograd_context_id`` and ``autograd_message_id`` to look up the 
+   appropriate ``send`` function.
+7. If this is the first time a worker has received a request for the given 
+   ``autograd_context_id``, it will compute dependencies locally as described 
+   in points 1-3 above.
+8. The ``send`` function retrieved in 6. is then enqueued for execution on the 
+   local autograd engine for that worker.
+9. Finally, instead of accumulating the gradients on the ``.grad`` field of the 
+   Tensor, we accumulate the gradients separately per 
+   `Distributed Autograd Context`_. The gradients are stored in a 
+   ``Dict[Tensor, Tensor]``, which is basically a map from Tensor to its 
+   associated gradient and this map can be retrieved using the 
+   ``get_gradients`` API.
```

Comment:
Suggestion: Given how many moving parts there are in this process, consider adding a diagram of this flow to visual readers.

Reply:
Got it